File: 102.二叉树的层序遍历.py
#
# @lc app=leetcode.cn id=102 lang=python3
#
# [102] 二叉树的层序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self):
        self.res  = []
        
    def dfs(self, root):
        if not root:return []
    
    # 暴力法
    def brute(self, root):
        # Time Limit Exceeded
        stack = [root]
        while stack and root:
            tmp = []
            _res = []
            for _ in stack:
                _res.append(_.val)
                print(_)
                if _.left:tmp.append(_.left)
                if _.right:tmp.append(_.right)
            stack = tmp
            self.res.append(_res)
        return self.res
            
        
        
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        return self.brute(root)
# @lc code=end

File: questions/102.二叉树的层序遍历.go
/*
 * @lc app=leetcode.cn id=102 lang=golang
 *
 * [102] 二叉树的层序遍历
 *
 * https://leetcode-cn.com/problems/binary-tree-level-order-traversal/description/
 *
 * algorithms
 * Medium (64.18%)
 * Likes:    806
 * Dislikes: 0
 * Total Accepted:    267.2K
 * Total Submissions: 416.2K
 * Testcase Example:  '[3,9,20,null,null,15,7]'
 *
 * 给你一个二叉树，请你返回其按 层序遍历 得到的节点值。 （即逐层地，从左到右访问所有节点）。
 *
 *
 *
 * 示例：
 * 二叉树：[3,9,20,null,null,15,7],
 *
 *
 * ⁠   3
 * ⁠  / \
 * ⁠ 9  20
 * ⁠   /  \
 * ⁠  15   7
 *
 *
 * 返回其层序遍历结果：
 *
 *
 * [
 * ⁠ [3],
 * ⁠ [9,20],
 * ⁠ [15,7]
 * ]
 *
 *
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	var res [][]int
	// 将根节点放在二维列表中
	q := []*TreeNode{root}
	// 遍历左右节点，左右兄弟节点放在同一层
	for len(q) > 0 {
		level := []int{}
		for _, x := range q {
			q = q[1:]
			if x.Left != nil {
				q = append(q, x.Left)
			}
			if x.Right != nil {
				q = append(q, x.Right)
			}
			level = append(level, x.Val)
		}
		res = append(res, level)
	}
	return res
}

// @lc code=end

